File: week6/credit.py
class Card:
    def __init__(self, valid=False, prefix=0, length=0):
        self.valid = valid
        self.prefix = prefix
        self.length = length

def checksum(num):
    count = 0
    temp_num = num

    # Get the length of the card number
    while temp_num != 0:
        count += 1
        temp_num //= 10

    # Store the card number in a list
    digits = [int(d) for d in str(num).zfill(count)]

    # Calculate the checksum
    sum_digits = 0
    for i in range(count - 2, -1, -2):
        sum_digits += sum_of_digits(digits[i] * 2)
    for i in range(count - 1, -1, -2):
        sum_digits += digits[i]

    # Determine if the card is valid
    valid = (sum_digits % 10 == 0)
    prefix = digits[0] * 10 + digits[1]
    return Card(valid, prefix, count)

def sum_of_digits(n):
    sum_digits = 0
    while n != 0:
        sum_digits += n % 10
        n //= 10
    return sum_digits

def main():
    num = int(input("Please type in your card number: "))

    main_card = checksum(num)
    length = main_card.length
    prefix = main_card.prefix
    valid = main_card.valid

    print("count is", length)
    if valid and ((prefix >= 34 and prefix <= 37) and length == 15):
        print("AMEX")
    elif valid and ((prefix >= 51 and prefix <= 55) and length == 16):
        print("MASTERCARD")
    elif valid and ((prefix >= 40 and prefix <= 49) and (length == 13 or length == 16)):
        print("VISA")
    else:
        print("INVALID")

if __name__ == "__main__":
    main()

File: week1/credit.c
#include <stdio.h>
#include <cs50.h>

typedef struct
{
    bool valid;
    int prefix;
    int len;
} Card;

Card checksum(long int num);
int sum_of_digits(int n);

int main(void)
{
    long int num = get_long("Please type in your card number: ");

    Card main_card = checksum(num);
    int len = main_card.len;
    int prefix = main_card.prefix;
    bool valid = main_card.valid;

    printf("count is %d", len);
    if (valid && ((prefix >= 34 && prefix <= 37) && len == 15))
    {
        printf("AMEX\n");
    }
    else if (valid && ((prefix >= 51 && prefix <= 55) && len == 16))
    {
        printf("MASTERCARD\n");
    }
    else if (valid && ((prefix >= 40 && prefix <= 49) && (len == 13 || len == 16)))
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }

}

Card checksum(long int num)
{
    Card currrent_card;
    
    int sum = 0;
    // Get the length of the card number
    int count = 0;
    for (long int tmp = num; tmp != 0; count++)
    {
        tmp /= 10;
    }

    // Store the card number in an array
    int digits[count];
    for (int i = count - 1; i >= 0; i--)
    {
        digits[i] = num % 10;
        num /= 10;
    }

    // Multiply every other digit by 2, starting with the number’s second-to-last digit, and then add those products’ digits together
    for (int i = count - 2; i >= 0; i -= 2)
    {
        sum += sum_of_digits((digits[i] * 2));
    }

    // Get the sum of digits that weren't multiplied by 2
    for (int i = count - 1; i >= 0; i -= 2)
    {
        sum += digits[i];
    }    

    // Test its last digit if its a valid card number
    if (sum % 10 == 0)
    {
        currrent_card.valid = true;
    }
    else
    {
        currrent_card.valid = false;
    }
    currrent_card.prefix = (digits[0] * 10) + digits[1];
    printf("%d\n", sum);
    currrent_card.len = count;
    return (currrent_card);
}

int sum_of_digits(int n)
{
    int sum = 0;
    while (n != 0)
    {
        sum += n % 10;
        n /= 10;
    }
    return (sum);
}